File: python3/alerts.py
#!/usr/bin/env python3
# Copyright 2020 UMU618.
#
# Authors: UMU618 <umu618@hotmail.com>
#
import base64
import hashlib
import hmac
import json
import ssl
import time
import urllib.request
import urllib.parse

URL_ROOT = 'https://oapi.dingtalk.com'


def get_sign(timestamp, secret):
    to_sign = str(timestamp) + '\n' + secret
    h = hmac.new(secret.encode('utf-8'), to_sign.encode('utf-8'),
                 digestmod=hashlib.sha256).digest()
    return base64.b64encode(h)


def send_dingtalk(token, secret, text):
    if not token or not text:
      print('Message:', text)
      return True

    url = URL_ROOT+'/robot/send?access_token='+token
    if secret:
      timestamp = round(time.time() * 1000)
      sign = get_sign(timestamp, secret)
      url += '&timestamp='+str(timestamp)+'&sign='+urllib.parse.quote(sign)
    headers = {'Content-Type': 'application/json; charset=utf-8'}
    body = bytes(json.dumps(
        {'msgtype': 'text', 'text': {'content': text}}), 'utf-8')
    try:
        req = urllib.request.Request(url, data=body, headers=headers)
        response = urllib.request.urlopen(
            req, context=ssl._create_unverified_context())
        if response.code != 200:
            print('fail to send dingtalk, status =', response.code)
            return False
        jo = json.loads(response.read())
        if jo['errcode'] == 0:
            print('Message sent:', text)
        else:
            print('fail to send dingtalk:', jo)
            return False
        return True
    except Exception as err:
        print('fail to send dingtalk, exception:', err)
        return False


if __name__ == '__main__':
    print('URL_ROOT:', URL_ROOT)
    secret = 'UMU'
    timestamp = 618
    sign = get_sign(timestamp, secret)
    print('secret:', secret)
    print('timestamp:', timestamp)
    print('sign:', sign.decode('utf-8'))

File: node.js/utils/alerts.js
/**
 * @author UMU618 <umu618@hotmail.com>
 * @copyright MEET.ONE 2019
 * @description Use block-always-using-brace npm-coding-style.
 * @lastModified 2020-04-08 14:37
 */

'use strict'

const fetch = require('node-fetch')

module.exports = {
  timeout: 0

  , setTimeout(t) {
    if (typeof(t) === 'number') {
      module.exports.timeout = t
    }
  }

  , getDingtalkSign: (timestamp, secret) => {
    const to_sign = timestamp + '\n' + secret
    const crypto = require('crypto')
    const hmac = crypto.createHmac('sha256', secret)
    return hmac.update(to_sign).digest('base64')
  }

  , sendDingtalk: (token, secret, text) => {
    if (!token || !text) {
      console.log('sendDingtalk() local message:', text)
      return
    }

    let url = 'https://oapi.dingtalk.com/robot/send?access_token=' + token
    if (secret) {
      const timestamp = Date.now()
      url += '&timestamp=' + timestamp + '&sign='
        + encodeURIComponent(module.exports.getDingtalkSign(timestamp, secret))
    }
    fetch(url, {
      method: 'POST'
      , headers: {
        'Content-Type': 'application/json'
      }
      , body: JSON.stringify({
        msgtype: 'text'
        , text: {
          content: text
        }
      })
      , timeout: module.exports.timeout
    }).then((res) => {
      if (res.ok) {
        return res.json()
      } else {
        console.error(
          `sendDingtalk() failed, status = ${res.status}, message: ${text}`)
        return null
      }
    }, (err) => {
      console.error(err)
    }).then((jo) => {
      if (jo !== null) {
        if (jo.errcode === 0) {
          console.log('sendDingtalk() succeeded:', text)
        } else {
          console.error(
            `sendDingtalk() failed: ${JSON.stringify(jo)}, message: ${text}`)
        }
      }
    })
  }

  , sendTelegram: (token, chat_id, text) => {
    if (!token || !chat_id || !text) {
      console.log('sendTelegram() local message:', text)
      return
    }
    fetch('https://api.telegram.org/bot' + token + '/sendMessage?'
      + require('querystring').stringify({
        chat_id: chat_id, text: text
      }), { method: 'POST', timeout: module.exports.timeout })
      .then(function (res) {
        if (res.ok) {
          return res.json()
        } else {
          console.error(
            `sendTelegram() failed, status = ${res.status}, message: ${text}`)
          return null
        }
      }, function (err) {
        console.error(err)
      })
      .then((jo) => {
        if (jo !== null) {
          if (jo.ok) {
            console.log('sendTelegram() succeeded:', text)
          } else {
            console.error(
              `sendTelegram() failed: ${JSON.stringify(jo)}, message: ${text}`)
          }
        }
      })
  }

  , sendFeishu: (token, text, title) => {
    if (!token || !text) {
      console.log('sendFeishu() local message:', text)
      return
    }
    const url = 'https://open.feishu.cn/open-apis/bot/hook/' + token
    const jo = {
      text: text
    }
    if (title) {
      jo.title = title
    }
    fetch(url, {
      method: 'POST'
      , headers: {
        'Content-Type': 'application/json'
      }
      , body: JSON.stringify(jo)
      , timeout: module.exports.timeout
    }).then((res) => {
      if (res.ok) {
        return res.json()
      } else {
        console.error(
          `sendFeishu() failed, status = ${res.status}, message: ${text}`)
        return null
      }
    }, (err) => {
      console.log(err)
    }).then((jo) => {
      if (jo !== null) {
        if (jo.ok === true) {
          console.log('sendFeishu() succeeded:', text)
        } else {
          console.error(
            `sendFeishu() failed: ${JSON.stringify(jo)}, message: ${text}`)
        }
      }
    })
  }
}
